File: Heap/Java/Minimum Cost to Hire K Workers.java
// Problem link: https://leetcode.com/problems/minimum-cost-to-hire-k-workers/description/
// Time: O(nlogn)
// Space: O(n)
class Solution {
    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        List<double[]> ratio = new ArrayList<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        double totalQuality = 0;
        double res = Double.POSITIVE_INFINITY;

        for (int i = 0; i < quality.length; i++) {
            ratio.add(new double[]{(double) wage[i] / quality[i], (double) quality[i]});
        }

        ratio.sort(Comparator.comparingDouble(a -> a[0]));

        for (double[] pair : ratio) {
            maxHeap.offer((int) pair[1]);
            totalQuality += pair[1];

            if (maxHeap.size() == k) {
                res = Math.min(res, pair[0] * totalQuality);
                totalQuality -= maxHeap.poll();
            }
        }

        return res;
    }
}

File: Heap/Python/Minimum Cost to Hire K Workers.py
# Problem link: https://leetcode.com/problems/minimum-cost-to-hire-k-workers/description/
# Time: O(nlogn)
# Space: O(n)
class Solution:
    def mincostToHireWorkers(self, quality: List[int], wage: List[int], k: int) -> float:
        ratio, maxHeap, total_quality, res = [], [], 0, inf

        for i in range(len(quality)):
            ratio.append((wage[i] / quality[i], quality[i]))
        
        ratio.sort()

        for rate, quality in ratio:
            heapq.heappush(maxHeap, -quality)
            total_quality += quality
            
            if len(maxHeap) == k:
                res = min(res, rate * total_quality)
                total_quality += heapq.heappop(maxHeap)

        return res
        
